Feature/graphql download queue subscription send only updates

schroda opened the pull request:
the old download changed subscription has been marked as deprecated and a new subscription has been added

the new subscription sends only changed downloads instead of the full status with every message

the type includes the following fields

- the `initial` download queue within the initial message **(n+1)**
- `updates` for
  - en- and dequeuing  **(n+1 without combination of omitting updates)**
  - repositioning downloads
  - errors
  - finished downloads
- an option `maxUpdates` to omit updates by setting a max update limit to prevent n+1 problem - includes a flag `omittedUpdates` which indicates that the `downloadStatus` queue should get re-fetched

schroda committed:
Emit only download changes instead of full status

The download subscription emitted the full download status, which, depending on how big the queue was, took forever because the graphql subscription does not support data loader batching, causing it to run into the n+1 problem

schroda committed:
Rename "DownloadManager#status" to "DownloadManager#updates"

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/manga/impl/download/DownloadManager.kt:
```diff
@@ -119,8 +126,8 @@ object DownloadManager {
-    private val statusFlow = MutableSharedFlow<DownloadStatus>()
-    val status = statusFlow.onStart { emit(getStatus()) }
+    private val updatesFlow = MutableSharedFlow<DownloadUpdates>()
+    val updates = updatesFlow.onStart { emit(getDownloadUpdates()) }
```
Shouldn't this be getStatus so it sends all current downloads the first time, so the client can cache and use them?
```suggestion
    val updates = updatesFlow.onStart { emit(getStatus()) }
```

schroda replied:
the initial subscription is one of the main reasons for this change, it causes the n+1 problem, because the subscription does not support data loader batching

to get the full status the download status query should be used and then after that, the subscription informs about changes

Robonau replied:
relying on two separate requests to link up perfectly seems like a bad idea

schroda replied:
I don't see how that is bad, the subscription is there to inform about changes and not to push the whole status everytime the progress of a single download has changed

you could even ignore the whole download queue and just use the updates from the subscription to show what is actively downloading

apollo is also using this as an example on when to use subscriptions https://www.apollographql.com/docs/react/data/subscriptions/#when-to-use-subscriptions -> Small, incremental changes to large objects

Robonau replied:
if the query responds with ```
[
    {id:1234, status:"running"},
    {id:1235, status:"idle"}
]```
but by the time the server deals with the incoming web-socket if 1234 has finished i'tl forever be stuck in the queue on the client

inversely if you do the web-socket first you don't know if the data from the web-socket or the query is more up to date if there is a difference

Robonau replied:
and if we are talking about only sending pieces when something changes then it would make sense to send the whole queue at first
since it is functionally changing from [] (an empty array) to the entire queue on the client
so the changes are the entire queue

Robonau replied:
> not to push the whole status everytime the progress of a single download has changed

i don't see why you are saying this? nobody as said we want this, we already have this currently.

what syer suggested and what would fix the issue i mentioned is having the one web-socket not a web-socket and a query.
 have that one web-socket deal with both the initial full list and the updates after the initialisation.
 if you really want the initial list to be optional you can make it a separate field in the gql
 
 ```gql
subscription MySubscription {
  newdownloadChanged {
    initial {
      state
      tries
      progress
      manga {
        id
      }
      chapter {
        id
      }
    }
    updates {
      state
      tries
      progress
      manga {
        id
      }
      chapter {
        id
      }
    }
  }
}
```

where initial is null in every message but the first

schroda replied:
> if the query responds with  [ {id:1234, status:"running"}, {id:1235, status:"idle"} ]
but by the time the server deals with the incoming web-socket if 1234 has finished i'tl forever be stuck in the queue on the client
inversely if you do the web-socket first you don't know if the data from the web-socket or the query is more up to date if there is a difference

which I'd say is a rare edge case with a relatively low chance of occurring

> and if we are talking about only sending pieces when something changes then it would make sense to send the whole queue at first
since it is functionally changing from [] (an empty array) to the entire queue on the client
so the changes are the entire queue

the changes are from the servers pov and not the clients?

> i don't see why you are saying this? nobody as said we want this, we already have this currently.

I don't get what you're saying here

> what syer suggested and what would fix the issue i mentioned is having the one web-socket not a web-socket and a query.

it's literally the main reason why I've made this change, the bigger the queue is the worse the responsiveness of the server gets, to literally making it unusable

> the initial subscription is one of the main reasons for this change, it causes the n+1 problem, because the subscription does not support data loader batching

I rather take some unlikely edge case for using a query + subscription compared to this

> if you really want the initial list to be optional

I don't want it at all, neither required nor optional - but I also don't care if something like this needs to be added

schroda committed:
Add initial queue to download subscription type

Adds the current queue at the time of sending the initial message.
This field is null for all following messages after the initial one

schroda committed:
Optionally limit and omit download updates

To prevent the n+1 dataloader issue, the max number of updates included in the download subscription can be limited.
This way, the problem will be circumvented and instead, the latest download status should be (re-)fetched via the download status query, which does not run into this problem.

schroda committed:
Formatting
